crates/context/config/src/client/protocol/stellar.rs

```diff
@@ -0,0 +1,392 @@
+                .try_into_val(&env)
+                .map_err(|_| StellarError::Custom {
+                    operation: ErrorOperation::Query,
+                    reason: "Failed to convert to Val".to_owned(),
+                })?;
+            let sc_val: ScVal = val.try_into_val(&env).map_err(|_| StellarError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Failed to convert to ScVal".to_owned(),
+            })?;
+            args = Some(vec![sc_val]);
+        }
+
+        match request.operation {
+            Operation::Read { method } => network.query(&contract, &method, args).await,
+            Operation::Write { method } => network.mutate(&contract, &method, args).await,
+        }
+    }
+}
+
+impl Network {
+    async fn query(
+        &self,
+        contract: &Contracts,
+        method: &str,
+        args: Option<Vec<ScVal>>,
+    ) -> Result<Vec<u8>, StellarError> {
+        let account = self
+            .client
+            .get_account(self.keypair.public_key().as_str())
+            .await
+            .map_err(|e| StellarError::Custom {
+                operation: ErrorOperation::Query,
+                reason: e.to_string(),
+            })?;
+
+        let source_account = Rc::new(RefCell::new(account));
+
+        let transaction = TransactionBuilder::new(source_account, self.network.as_str(), None)
+            .fee(10000u32)
+            .add_operation(contract.call(method, args))
+            .set_timeout(15)
+            .expect("Transaction timeout")
+            .build();
+
+        let result: Result<RawSimulateTransactionResponse, Error> = self
+            .client
+            .simulate_transaction(transaction.clone(), None)
+            .await;
+        let xdr_results: Vec<RawSimulateHostFunctionResult> = result.unwrap().results.unwrap();
+
+        match xdr_results.first().and_then(|xdr| xdr.xdr.as_ref()) {
+            Some(xdr_bytes) => {
+                let xdr_bytes = base64::engine::general_purpose::STANDARD
+                    .decode(xdr_bytes)
+                    .map_err(|_| StellarError::Custom {
+                        operation: ErrorOperation::Query,
+                        reason: "Failed to decode XDR response".to_owned(),
+                    })?;
+
+                let cursor = Cursor::new(xdr_bytes);
+                let mut limited = xdr::Limited::new(cursor, xdr::Limits::none());
+                match ScVal::read_xdr(&mut limited) {
+                    Ok(ScVal::Bytes(bytes)) => Ok(bytes.into()),
+                    Ok(_) => Err(StellarError::Custom {
+                        operation: ErrorOperation::Query,
+                        reason: "Unexpected XDR response type; expected ScVal::Bytes".to_owned(),
+                    }),
+                    Err(_) => Err(StellarError::Custom {
+                        operation: ErrorOperation::Query,
+                        reason: "Failed to parse XDR type; expected ScVal::Bytes".to_owned(),
```

Comment:
why are we only expecting `ScVal::Bytes`? we want to encode everything we get (whatever it may be) to bytes.

As we already know, `ToXdr` helps with something like this.

Reply:
Updated to handle all sc_val types